### DISCUSSION TITLE
[Proposal] Provide a way to opt-in specified optional parameters decorated with CallerMemberName, CallerArgumentExpression, CallerFilePath, CallerLineNumber as a warning or error

### DISCUSSION THREAD

**Discussion by raramer** [dotnet/csharplang#8382] [Language Ideas] [answered]
Most of the time when using `CallerMemberNameAttribute`, `CallerArgumentExpressionAttribute`, `CallerFilePathAttribute`, or `CallerLineNumberAttribute` I don't want or expect the calling code to actually specify the optional value, and instead want the compiler to generate the value(s).  I would like a built-in way (i.e. not a custom code analyzer) to opt-in the parameters such that the compiler would issue a warning or error if calling code does specify a value.

For example:
```
public void DoProcessing()
{
    TraceMessage("Something happened.", "CustomValue"); // would like this to optionally show as a warning or error
}

public void TraceMessage(string message, [System.Runtime.CompilerServices.CallerMemberName] string memberName = "")
{
    System.Diagnostics.Trace.WriteLine($"{memberName}: {message});
}
```

**Options** 
1. Add isError / isWarning options to `CallerMemberNameAttribute`, `CallerArgumentExpressionAttribute`, `CallerFilePathAttribute`, and `CallerLineNumberAttribute` to set the level of reporting, similar to how ObsoleteAttribute works.
2. Update ObsoleteAttribute to allow setting on parameters.  This sounds like the easiest, but I'm sure there's some hidden caveats, as it's not currently supported.
3. Add an ObsoleteParameterAttribute that works like ObsoleteAttribute, but applies just to parameters.

**Comment by CyrusNajmabadi** [comment, marked as answer]
When you want to report an error on code based on presence of an attribute, the best way to do that is through an analyzer.  This woudl be a very simple one to write, where you just analyzer IArgumentOperations, see if they are mapping to a parameter which has such an attribute, and if hte argument is implicit or not.
